src/krbd.cc

```diff
@@ -241,60 +243,78 @@ static int wait_for_udev_add(struct udev_monitor *mon, const krbd_spec& spec,
   for (;;) {
     struct pollfd fds[1];
     struct udev_device *dev;
-    int r;
 
     fds[0].fd = udev_monitor_get_fd(mon);
     fds[0].events = POLLIN;
     r = poll(fds, 1, POLL_TIMEOUT);
-    if (r < 0)
-      return -errno;
+    if (r <= 0) {
+      r = (r == 0) ? -ETIMEDOUT : -errno;
+      break;
+    }
 
-    if (r == 0)
-      return -ETIMEDOUT;
 
     dev = udev_monitor_receive_device(mon);
     if (!dev)
       continue;
 
-    if (strcmp(udev_device_get_action(dev), "add") != 0)
-      goto next;
+    if (strcmp(udev_device_get_action(dev), "add") != 0) {
+      udev_device_unref(dev);
+      continue;
+    }
 
-    if (!bus_dev) {
-      if (strcmp(udev_device_get_subsystem(dev), "rbd") == 0) {
+    const char* sub_sys = udev_device_get_subsystem(dev);
+    if (strcmp(sub_sys, "rbd") == 0) {
+      if (!bus_dev) {
         auto cur_spec = spec_from_dev(dev);
         if (cur_spec && *cur_spec == spec) {
           bus_dev = dev;
-          continue;
+          goto check;
         }
       }
-    } else {
-      if (strcmp(udev_device_get_subsystem(dev), "block") == 0) {
-        const char *major = udev_device_get_sysattr_value(bus_dev, "major");
-        const char *minor = udev_device_get_sysattr_value(bus_dev, "minor");
-        const char *this_major = udev_device_get_property_value(dev, "MAJOR");
-        const char *this_minor = udev_device_get_property_value(dev, "MINOR");
+    } else if (strcmp(sub_sys, "block") == 0) {
+      block_dev_list.push_back(dev);
+      goto check;
+    }
+
+    udev_device_unref(dev);
+
+check:
+    if (bus_dev && !block_dev_list.empty()) {
+      const char *major = udev_device_get_sysattr_value(bus_dev, "major");
+      const char *minor = udev_device_get_sysattr_value(bus_dev, "minor");
+      assert(!minor ^ have_minor_attr());
 
-        ceph_assert(!minor ^ have_minor_attr());
+      auto p = block_dev_list.begin();
+      while (p != block_dev_list.end()) {
+        const char *this_major = udev_device_get_property_value(*p, "MAJOR");
+        const char *this_minor = udev_device_get_property_value(*p, "MINOR");
 
         if (strcmp(this_major, major) == 0 &&
             (!minor || strcmp(this_minor, minor) == 0)) {
           string name = get_kernel_rbd_name(udev_device_get_sysname(bus_dev));
 
-          ceph_assert(strcmp(udev_device_get_devnode(dev), name.c_str()) == 0);
+          assert(strcmp(udev_device_get_devnode(*p), name.c_str()) == 0);
           *pname = name;
-
-          udev_device_unref(dev);
-          udev_device_unref(bus_dev);
-          break;
+          goto done;
+        } else {
+          udev_device_unref(*p);
+          p = block_dev_list.erase(p);
```

Comment:
I don't think erasing-as-we-go is worth it.  This vector won't have more than a few items, all of them can be taken care of at the bottom.  Let's change this to a simple for loop.